algebird-core/src/main/scala/com/twitter/algebird/experimental/BloomFilter.scala

```diff
@@ -0,0 +1,413 @@
+          else
+            scala.math.pow(1 - scala.math.exp(-numHashes * size.estimate * 1.1 / width), numHashes)
+
+        ApproximateBoolean(true, 1 - fpProb)
+      } else {
+        // False negatives are not possible.
+        ApproximateBoolean.exactFalse
+      }
+
+    /**
+     * This may be faster if you don't care about evaluating
+     * the false positive probability
+     */
+    def maybeContains(item: A): Boolean
+
+    // Estimates the cardinality of the set of elements that have been
+    // inserted into the Bloom Filter.
+    def size: Approximate[Long]
+
+    def toBitSet: BitSet
+
+    /**
+     * Compute the Hamming distance between the two Bloom filters
+     * `a` and `b`. The distance is defined as the number of bits that
+     * need to change to in order to transform one filter into the other.
+     */
+    def hammingDistance(that: Hash): Int =
+      (this, that) match {
+        // Comparing with empty filter should give number
+        // of bits in other set
+        case (Empty, Empty)   => 0
+        case (Empty, y: Hash) => y.numBits
+        case (x: Hash, Empty) => x.numBits
+
+        // Otherwise compare as bit sets
+        case (_, _) => (this.toBitSet ^ that.toBitSet).size.toInt
+      }
+
+  }
+
+  case object Empty extends Hash {
+    override def toBitSet: BitSet = BitSet.empty
+
+    override val numBits = 0
+
+    override def ++(other: Hash): Hash = other
+
+    override def +(other: A): Item = Item(other)
+
+    override def checkAndAdd(other: A): (Hash, ApproximateBoolean) =
+      (this + other, ApproximateBoolean.exactFalse)
+
+    override def contains(item: A): ApproximateBoolean = ApproximateBoolean.exactFalse
+
+    override def maybeContains(item: A): Boolean = false
+
+    override def size: Approximate[Long] = Approximate.exact[Long](0)
+  }
+
+  case class Item(item: A) extends Hash {
+    override val numBits: Int = numHashes
+
+    override def toBitSet: BitSet = BitSet(hashToArray(item))
+
+    override def ++(other: Hash): Hash =
+      other match {
+        case Empty => this
+        case Item(otherItem) =>
+          Instance {
+            BitSet
+              .newEmpty(0)
+              .mutableAdd(hashToArray(item))
+              .mutableAdd(hashToArray(otherItem))
+          }
+        case _ => other + item
+      }
+
+    override def +(other: A): Hash = this ++ Item(other)
+
+    override def checkAndAdd(other: A): (Hash, ApproximateBoolean) =
+      if (other == item) {
+        (this, ApproximateBoolean.exactTrue)
+      } else {
+        (this + other, ApproximateBoolean.exactFalse)
+      }
+
+    override def contains(x: A): ApproximateBoolean = ApproximateBoolean.exact(item == x)
+
+    override def maybeContains(x: A): Boolean = item == x
+
+    override def size: Approximate[Long] = Approximate.exact[Long](1)
+  }
+
+  /*
+   * Bloom filter with multiple values
+   */
+  case class Instance(bits: BitSet) extends Hash {
```

Comment:
I think this is the only subclass we need here. The point of BitSet is that it is very fast. We don't need to allocate for an empty one.

We could have two cases, Item and Instance, and use Item to defer hashing until hopefully we have a number of them to combine at once.

on the other hand, there is a nice invariant here that this is always two or more items, we could even check that law in tests... though I guess if we merge an item with itself we do get a single element bitset.

maybe leave ignore all this and just update the comment to explain this has 1 or more bit set, and maybe we can check that property.